uaclient/log.py

```diff
@@ -0,0 +1,59 @@
+import json
+import logging
+import re
+from collections import OrderedDict
+from typing import Any, Dict, List
+
+
+class JsonArraryFormatter(logging.Formatter):
+    default_time_format = "%Y-%m-%dT%H:%M:%S%z"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not self.datefmt:
+            self.datefmt = self.default_time_format
+        self._required_fields = self.get_required_fields()
+
+    def get_required_fields(self) -> List[str]:
+        # We only support the % style formatting...currently
+        if isinstance(self._style, logging.PercentStyle):
+            formatter_style_pattern = re.compile(r"%\((.+?)\)", re.IGNORECASE)
+        else:
+            raise ValueError("Invalid format: %s" % self._fmt)
+
+        if self._fmt:
+            return formatter_style_pattern.findall(self._fmt)
+        else:
+            return []
+
+    def update_local_log_records(
+        self,
+        record: logging.LogRecord,
+        local_log_record: Dict[str, Any],
+        message_dict: Dict[str, Any],
+    ):
+        # update the required fields in the order stated
+        for field in self._required_fields:
+            local_log_record[field] = record.__dict__.get(field)
+
+        # update the exc and stack info
+        local_log_record.update(message_dict)
```

Comment:
The message_dict contains message-specific fields that don't occur in all messages right?
In that case, I think it should be it's own field and formatted as a json object.

For example:
```json
["EXCEPTION", "datetime",  "here is the actual log message", {"exc_info": "stuff here", "stack_info": "lots\nof\nlines\nhere"}]
```
Note that my example doesn't include all the fields that we should include - it only illustrates how I think the message_dict shuold get included - see the last item in the example.

Reply:
Okay, so that it is in line with how we handle the `extra` field.

Reply:
I think I will move the `extra` field to the message_dict, as well since it is not always there.

Reply:
Here is the difference with the optional extra vs having extra always https://paste.ubuntu.com/p/34tSJmb79C/